[Coupons] Product selector list

0nko opened the pull request:
This PR is the first part of a product selector dialog (it's already too big, sorry about that), which is used in a conditional application of coupons to specific products.

Right now, only a list of products is displayed, with the products that are included in the coupon preselected. You can change the selection and the button title is updated, but it doesn't do anything yet. The search and filtering has also not been implemented yet.

**To test:**
1. Go to coupons, tap on one and go to the edit screen
2. Notice there are 2 new buttons, under the `Apply coupon to`: `All products ` and `Select (or Edit) products`
3. Currently, both just open the product selection
4. Verify the product list is loaded properly

Note the only regular products (not variations) are currently supported for selection.

![image](https://user-images.githubusercontent.com/1522856/169794772-c4a9e4a1-423f-4f74-af51-aa21143c170d.png)
![image](https://user-images.githubusercontent.com/1522856/169795520-6c7f02f4-012c-42c7-bffb-e49fca2136f1.png)

0nko committed:
Add custom rounded checkmark icons

0nko committed:
Add string and dimens resources

0nko committed:
Add a simplified function to include plurals string resource in Compose

0nko committed:
Add product selector VM, fragment, UI and data handling

0nko committed:
Update FluxC reference

0nko committed:
Add navigation from the edit screen to the product selector

0nko committed:
Update the button paddings and titles

0nko committed:
Pre-select included products

0nko committed:
Change the missing image icon

0nko committed:
Fix detekt issues

0nko committed:
Merge branch 'issue/5537-product-restriction' into issue/5537-product-restriction-list

# Conflicts:
#	WooCommerce/src/main/res/values/strings.xml
#	build.gradle

hafizrahman reviewed (commented):
This doesn't specifically need to be on this PR but I wonder if we should make the header area smaller here on landscape mode:

![Screen Shot 2022-05-24 at 14 35 12](https://user-images.githubusercontent.com/266376/169975263-96365d35-8f0a-499b-a785-04cac23cdcd2.png)

Only being able to scroll 2 items at a time does not feel like a comfortable experience.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/Utils.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Composable
+fun pluralsResource(@PluralsRes id: Int, quantity: Int) =
+    LocalContext.current.resources.getQuantityString(id, quantity, quantity)
```
`plurals` are not supported by Glotpress, we should instead use our own `getQuantityString` that's defined [here](https://github.com/woocommerce/woocommerce-android/blob/03ee93169d9ef1082061ae6e761f50ef52899972/WooCommerce/src/main/kotlin/com/woocommerce/android/util/StringUtils.kt#L43-L55).

0nko replied:
Right 👍. Thanks!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EditCouponNavigator.kt:
```diff
@@ -4,21 +4,31 @@ import androidx.fragment.app.Fragment
         when (target) {
             is OpenDescriptionEditor -> {
-                navController.navigate(
+                navController.navigateSafely(
```
Thanks for fixing this 👍.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EditCouponScreen.kt:
```diff
@@ -173,14 +177,59 @@ private fun DetailsSection(
+            style = MaterialTheme.typography.body2,
+            color = colorResource(id = R.color.color_on_surface_medium)
+        )
+        WCOutlinedButton(
+            onClick = onAllProductsButtonClick,
+            text = stringResource(id = R.string.coupon_conditions_products_all_products_button_title),
+            colors = ButtonDefaults.outlinedButtonColors(contentColor = MaterialTheme.colors.onSurface),
+            modifier = Modifier.fillMaxWidth()
+        )
```
I think it was decided to remove the "All products" button, I remember Huong mentioning it before, but I don't remember where, but you can check the updated designs in Figma in the iOS page (for information, the Android page seems to be outdated a bit).

0nko replied:
Thanks, I also noticed this after testing the iOS app 👍

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductRepository.kt:
```diff
@@ -0,0 +1,72 @@
+import org.wordpress.android.fluxc.store.WCProductStore
+import javax.inject.Inject
+
+class ProductRepository @Inject constructor(
```
Just a question, since we have a different `ProductListRepository` in the same package, do you think we should change the name, or at least add a comment to this class to explain that it's a new version that uses coroutines instead of EventBus?

0nko replied:
Makes sense 👍I've thought about merging the 2 repositories, but in the end I just found it cleaner to rename (`ProductSelectorRepository`) and move it to the selector package.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -0,0 +1,402 @@
+) {
+    Row(
+        modifier = Modifier
+            .clickable(
```
np: maybe add `onClickLabel` for accessibility?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorFragment.kt:
```diff
@@ -0,0 +1,45 @@
+            navigationIcon = R.drawable.ic_gridicons_cross_24dp
+        )
+
+    private var screenTitle: String by observable("") { _, oldValue, newValue ->
+        if (oldValue != newValue) {
+            updateActivityTitle()
+        }
+    }
```
np, I don't think this is needed, the screen title seems to be static.

0nko replied:
Oops, that's a leftover from the copied fragment.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -0,0 +1,402 @@
+            }
+
+            if (!product.sku.isNullOrEmpty()) {
+                ProductListItemInfo("SKU: ${product.sku}")
```
np: should this be translatable string? There's something similar in `R.string.orderdetail_product_lineitem_sku_value`

0nko replied:
Good catch 👍

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -0,0 +1,402 @@
+            .fillMaxHeight()
+            .background(color = MaterialTheme.colors.surface)
+    ) {
+        Text(
+            text = "Select products",
+            style = MaterialTheme.typography.h4,
+            color = MaterialTheme.colors.onSurface,
+            fontWeight = FontWeight.Bold,
+            modifier = Modifier
+                .padding(dimensionResource(id = R.dimen.major_100))
+        )
```
Just sharing this discussion that I had with Joe regarding the large titles we have in the designs, they are supposed to be collapsible, and we agreed that we should keep using the regular titles for now until we have enough time to work on them, and have an implementation that works for the whole app (https://www.figma.com/file/O8MT8FHx8N2zP9pc2namgl?node-id=1469:11959#186324308), so maybe we can remove this from here, and just set the screentitle on the Fragment itself, WDYT?

0nko replied:
Thank you, I missed that!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorViewModel.kt:
```diff
@@ -0,0 +1,194 @@
+        .asLiveData()
+
+    init {
+        if (searchQuery.value == null) {
+            viewModelScope.launch {
+                isLoading.value = true
+                productListHandler.fetchProducts(forceRefresh = true)
+                isLoading.value = false
+            }
+        }
```
Just sharing something: later on, I found out that this logic causes `fetchX` to be called twice, since we are calling it again in `monitorSearchQuery`.
I've fixed the behavior in #6575 [here](https://github.com/woocommerce/woocommerce-android/blob/4afc55b8669b2c62a2328546901bfac7f3a86962/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/CouponListViewModel.kt#L133-L135), if you want to fix it too, either on this PR, or next PR when you implement search.

0nko replied:
Oh yeah, I remember that. I'll get rid of the search code for now as it's just polluting the PR and I'll added the fixed code later when I implement the search.

hichamboushaba reviewed (commented):
Really nice job here @0nko, I left some comments, most of them minor, but require your attention.

0nko committed:
Merge branch 'issue/5537-product-restriction' into issue/5537-product-restriction-list

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EditCouponScreen.kt

0nko committed:
Move list handler and repository to the selector package

0nko committed:
Remove the All products button

0nko committed:
Update the string resources

0nko committed:
Refactor the list item composable into a more generic one

0nko committed:
Remove the unneeded title update code

0nko committed:
Refactor the SelectionState out of the VM

0nko committed:
Remove unneeded opt-in

0nko committed:
Set a static screen title

0nko committed:
Remove search-related code

0nko committed:
Add loading-more indicator

0nko committed:
Optimize imports

0nko commented:
Thank you both for a thorough review! I hope I didn't miss anything but I think I've addressed all of your comments and the PR is ready for another round.

Note: I've refactored the list item Composable to be more generic, so that it can be reused for variations & I added a load-more indicator to the list.

0nko committed:
Use product-specific resources

0nko committed:
Count the selected products properly, including variations

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/edit/EditCouponNavigationTarget.kt:
```diff
@@ -10,4 +10,5 @@ sealed class EditCouponNavigationTarget : Event() {
         val currencyCode: String,
         val showLimitUsageToXItems: Boolean
     ) : EditCouponNavigationTarget()
+    data class NavigateToProductSelector(val selectedProductIds: List<Long>) : EditCouponNavigationTarget()
```
np, since we will have a different event to navigation to the product selector to select excluded products, WDYT about making it clear that this is for including products? For the categories selector, I selected the name `EditIncludedProductCategories` 🤔

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/selector/ProductSelectorScreen.kt:
```diff
@@ -0,0 +1,279 @@
+            thickness = dimensionResource(id = R.dimen.minor_10)
+        )
+
+        WCColoredButton(
+            onClick = { /*TODO*/ },
+            text = StringUtils.getQuantityString(
+                quantity = state.selectedProductCount,
+                default = R.string.product_selector_select_button_title_default,
+                one = R.string.product_selector_select_button_title_one
+            ),
+            modifier = Modifier
+                .fillMaxWidth()
+                .padding(dimensionResource(id = R.dimen.major_100)),
+            enabled = state.selectedProductCount > 0
+        )
```
Two points about this:
1. the button should be enabled always.
2. it should say "Done" when no product is selected.2. 

internal discussion: p1650938526831689/1649411442.609089-slack-C02TD63FZA8

0nko replied:
Yes, this is already addressed in #6600.

hichamboushaba reviewed (approved):
Nice work @0nko, I left two comments, but feel free to address them on next PRs.

0nko committed:
Rename the navigation target data class

hichamboushaba commented:
@0nko I noticed this PR was merged to a feature branch, was there a reason for this?
I believe merging to `trunk` directly would help us stay in sync better, as the buttons for categories and products are on the same page, and since everything is behind a feature flag, IMO there is no need to keep things outside of `trunk`.
WDYT?

0nko commented:
@hichamboushaba I kept it in a feature branch because the feature was non-functional/incomplete, but you're right. I'll open a new PR to merge it to `trunk`.

hichamboushaba commented:
Thanks so much @0nko, I'll update my PR #6602 to put the button on the correct spot after we merge your PR.
